packages/remix-validated-form/src/ValidatedForm.tsx

```diff
@@ -272,9 +275,11 @@ export function ValidatedForm<DataType>({
       ref={mergeRefs([formRef, formRefProp])}
       {...rest}
       action={action}
-      onSubmit={(event) => {
+      onSubmit={async (event) => {
```

Comment:
Making this async isn't enough because it's a synchronous callback. While this is awaiting validations, remix will just submit anyway. We'll need something like this:

```ts
const submit = useSubmit();

// ...
onSubmit={async event => {
  event.preventDefault();
  // everything else in here
  submit(/* the correct element */);
}}
```

Remix's `Form` does some shennanigans to get the correct element for the `submit` call ([FormImpl](https://github.com/remix-run/remix/blob/81e23269d89567fc8bf1eb9e43ea9f753ed59189/packages/remix-react/components.tsx#L835-L919)).

Reply:
This explains the test failures I assume! I was sure there would be a quirk to fix still

Reply:
@airjp73  on this, a common pattern is just a 2 step submit, first submit prevents default, validates and marks state us validated, then triggers a second form submission, that behaves normally like:

```ts
const [ hasValidated, setValidated ] = useState(false)

const handleSubmit = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      if (!hasValidated) {
        event.preventDefault();
        //Validation and what not
        setHasValidated(true); 
        formRef.current?.submit();
      } else {
        setHasValidated(false);
      }
    },
    [hasValidated, setHasValidated]
  );
```

Reply:
Submitting the form directly won't cut it, I'm afraid. From a comment in the `remix` source code:

```
When calling `submit` on the form element itself, we don't get data from
the button that submitted the event. For example:

<Form>
  <button name="something" value="whatever">Submit</button>
</Form>

formData.get("something") should be "whatever", but we don't get that
unless we call submit on the clicked button itself.

To figure out which button triggered the submit, we'll attach a click
event listener to the form. The click event is always triggered before
the submit event (even when submitting via keyboard when focused on
another form field, yeeeeet) so we should have access to that button's
data for use in the submit handler.
```

Since this is submission is now asynchronous, the user could click something else between the first submit and second submit, causing remix's button ref to change. This means we're probably going to have to do the same click handler trick that remix does.

If we have to do the extra work to detect the button anyway, I'd rather go for the `useSubmit` approach over a double submit. We'll also need to take into account that the form could be a fetcher form. Something like:
```ts
if (fetcher) fetcher.submit();
else submit(formRef.current);
```